### PR TITLE
Add `spmm` bf16 support

### PR DESCRIPTION
**Pull request by yanbing-j**
This PR is to add BF16 support for spmm kernel implementation in torch_sparse. And the dispatch name is needed to keep unique.
BF16 UT has been added in test_matmul.py, `matmul` can run into spmm kernel.

### CODE REVIEW

**Review by rusty1s — APPROVED** [review]

**Comment by yanbing-j** [comment]
Hi @rusty1s , according to the CI failures, I also add BF16 support for `spspmm`, and update corresponding UTs. These test cases are based on torch_scatter bf16. So torch_scatter bf16 needs to be imported into torch_sparse CI check.

**Comment by rusty1s** [comment]
I see :( Shall we guide the `bf16` test on whether `torch-scatter>2.0.9` is used?

**Comment by yanbing-j** [comment]
Yes. I add `torch.bfloat16` only when `torch_scatter.__version__ > 2.0.9`. So current version will not run into bf16 test cases. I test this in my local, bf16 test cases can pass.

File: csrc/cpu/spspmm_cpu.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ spspmm_cpu(torch::Tensor rowptrA, torch::Tensor colA,
             if (HAS_VALUE)
               tmp_vals[cB] += valA_data[eA] * valB_data[eB];
             else
-              tmp_vals[cB]++;
+              tmp_vals[cB] += 1;
```
Comment (rusty1s) [inline csrc/cpu/spspmm_cpu.cpp:80]: What's the reason for this change?

File: test/test_spspmm.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def test_sparse_tensor_spspmm(dtype, device):
         ], dtype=dtype, device=device),
     )
 
-    expected = torch.eye(10, dtype=dtype, device=device)
+    expected = torch.eye(10, device=device).to(dtype)
```
Comment (rusty1s) [inline test/test_spspmm.py:38]: What's the reason for this change?

File: test/utils.py [hunk 3]
```diff
@@ -5,6 +5,12 @@
 dtypes = [torch.half, torch.float, torch.double, torch.int, torch.long]
 grad_dtypes = [torch.half, torch.float, torch.double]
 
+import torch_scatter
+if torch_scatter.__version__ > '2.0.9':
```
Comment (rusty1s) [inline test/utils.py:9]: I don't think you can use `>` like this in strings. You would test against the individual integers.

File: csrc/cpu/spspmm_cpu.cpp (hunk 1 above)
Comment (yanbing-j) [reply csrc/cpu/spspmm_cpu.cpp:80]: The runtime error shows `++` not support BFloat16 implementation.

File: test/test_spspmm.py (hunk 2 above)
Comment (yanbing-j) [reply test/test_spspmm.py:38]: `eye` doesn't have BFloat16 implementation.

File: test/utils.py (hunk 3 above)
Comment (yanbing-j) [reply test/utils.py:9]: This is similar with https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/profile/profiler.py#L107.

Comment (rusty1s) [reply test/utils.py:9]: I think then this is bugged as well. You can see that it breaks here:
```
'2.0.10' > '2.0.9'
```

Comment (yanbing-j) [reply test/utils.py]: Use `version.parse(torch_scatter.__version__) > version.parse("2.0.9")` can help. This is from PyTorch.

File: test/utils.py [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
 
 reductions = ['sum', 'add', 'mean', 'min', 'max']
 
 dtypes = [torch.half, torch.float, torch.double, torch.int, torch.long]
 grad_dtypes = [torch.half, torch.float, torch.double]
 
+if version.parse(torch_scatter.__version__) > version.parse("2.0.9"):
+    dtypes.append(torch.bfloat16)
+    grad_dtypes.append(torch.bfloat16)
+del torch_scatter
```
Comment (rusty1s) [inline test/utils.py:13]:
```suggestion
```

File: test/utils.py (hunk 3 above)
Comment (rusty1s) [reply test/utils.py]: Cool, thanks!
